head.go

```diff
@@ -1749,17 +1749,22 @@ func (s *memSeries) iterator(id int) chunkenc.Iterator {
 	}
 
 	if id-s.firstChunkID < len(s.chunks)-1 {
-		return c.chunk.Iterator()
+		return c.chunk.Iterator(it)
 	}
 	// Serve the last 4 samples for the last chunk from the sample buffer
 	// as their compressed bytes may be mutated by added samples.
-	it := &memSafeIterator{
-		Iterator: c.chunk.Iterator(),
+	if msIter, ok := it.(*memSafeIterator); ok {
+		msIter.Iterator = c.chunk.Iterator(nil)
+		msIter.i = -1
+		msIter.total = c.chunk.NumSamples()
+		msIter.buf = s.sampleBuf
```

Comment:
Shouldn't this return here?

Reply:
Yes, thanks, fixed.
Also made `msIter.Iterator = c.chunk.Iterator(nil)` -> `msIter.Iterator = c.chunk.Iterator(msIter.Iterator)`.